File: src/leetcode/2025/12/Solution219.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution219
{
public:
    bool containsNearbyDuplicate(vector<int> &nums, int k)
    {
        unordered_map<int, int> indexMap;

        for (int i = 0; i < nums.size(); i++)
        {
            if (indexMap.find(nums[i]) != indexMap.end())
            {
                if (i - indexMap[nums[i]] <= k)
                {
                    return true;
                }
            }
            indexMap[nums[i]] = i;
        }

        return false;
    }
};

int main()
{
    Solution219 solution;
    vector<int> nums1 = {1, 2, 3, 1};
    int k1 = 3;
    bool result1 = solution.containsNearbyDuplicate(nums1, k1);
    printf("%s\n", result1 ? "true" : "false");
    // Expected output: true

    vector<int> nums2 = {1, 0, 1, 1};
    int k2 = 1;
    bool result2 = solution.containsNearbyDuplicate(nums2, k2);
    printf("%s\n", result2 ? "true" : "false");
    // Expected output: true

    vector<int> nums3 = {1, 2, 3, 1, 2, 3};
    int k3 = 2;
    bool result3 = solution.containsNearbyDuplicate(nums3, k3);
    printf("%s\n", result3 ? "true" : "false");
    // Expected output: false

    return 0;
}

File: src/leetcode/2025/12/Solution219.java
public class Solution219 {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        java.util.HashSet<Integer> set = new java.util.HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            if (set.contains(nums[i])) {
                return true;
            }
            set.add(nums[i]);
            if (set.size() > k) {
                // Remove the oldest number to maintain the window size of k
                set.remove(nums[i - k]);
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Solution219 solution = new Solution219();

        // Example test cases
        int[] nums1 = { 1, 2, 3, 1 };
        int k1 = 3;
        System.out.println(solution.containsNearbyDuplicate(nums1, k1)); // true

        int[] nums2 = { 1, 0, 1, 1 };
        int k2 = 1;
        System.out.println(solution.containsNearbyDuplicate(nums2, k2)); // true

        int[] nums3 = { 1, 2, 3, 1, 2, 3 };
        int k3 = 2;
        System.out.println(solution.containsNearbyDuplicate(nums3, k3)); // false
    }
}
